### ISSUE TITLE
`space-before-function-paren` false positive

### ISSUE THREAD

**Issue by iam4x** [eslint/eslint#4442]
With rule: `'space-before-function-paren': [2, 'never'],` and using `babel-eslint@5.x` this is a false positive:

`export default async function({ flux, history, location }) {`

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
You're using an experimental syntax that ESLint doesn't natively support. Please file an issue on babel-eslint. cc @hzoo

**Comment by arv** [comment]
@hzoo Do you know when this changed? My build started breaking due to this as well.

**Comment by jkrems** [comment]
Afaict a possible fix is the following:

```js
        if (node.generator || node.async) { // async shifts the token offsets
            if (node.id) {
                leftToken = tokens[2];
                rightToken = tokens[3];
            } else {
                // Object methods are named but don't have an id
                leftToken = context.getTokenBefore(node);
                rightToken = tokens[0];
            }
```

Which doesn't explain why it broke though.

**Comment by hzoo** [comment]
This is basically because of https://github.com/babel/acorn-to-esprima/commit/482350ca852f2b5cd5007212c535424ad623c118 as a result of https://github.com/eslint/eslint/issues/4415#issuecomment-156594730.

I left it in there before since I figured it would break other rules. We can really only fix this by adding rules in https://github.com/babel/eslint-plugin-babel.

**Comment by mysticatea** [comment]
I think a side effect of #4428 fixes this.
